app/models/concerns/with_assignments_batch.rb

```diff
@@ -0,0 +1,29 @@
+# WithAssignmentsBatch mirrors the WithAssignment mixin
+# but implements operations in batches, so that they outperform
+# their counterparts
+module WithAssignmentsBatch
+  extend ActiveSupport::Concern
+
+  def find_assignments_for(user, _organization = Organization.current)
+    exercises = self.exercises
```

Comment:
doesn't the attribute get cache'd?

Reply:
I guess that because of how he uses it, it is preloading it on exercises object and not into guide itself. So it needs to assign it to a variable in order to access through that preload in the last line.

Reply:
Exactly. You need to preload a given instance of the relation object. If the relation instance changes, the preloading gets lost. 

I have actually seen samples that go even further and convert the relation into an array before doing this.

Reply:
Long story short, I was wrong. As @julian-berbel said, the collection proxy get cached, which means that this implementation had side-effects :open_mouth: 

I have re-implemented this method in a less obscure but more algorithmic way, which ensures no side effect is produced. Please review it again :pray: